File: wrappers/uniffi/kotlin/src/test/kotlin/com/equs/VcCoreTest.kt
package com.equs

import com.equs.credentials.*
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.Dispatcher
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okhttp3.mockwebserver.RecordedRequest
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class VcCoreTest {
    companion object {
        lateinit var mockServer: MockWebServer
        lateinit var keyMetadata: KeyMetadata
        lateinit var sharedKms: InMemKms
        lateinit var statusListJwt: String

        @JvmStatic
        @BeforeAll
        fun setup() {
            setJniLibPath()
            mockServer = MockWebServer().apply { start(9001) }
            sharedKms = InMemKms()
            keyMetadata =
                runBlocking { createDidAndKeyMetadata(sharedKms).keyMetadata }
            statusListJwt =
                runBlocking {
                    val statusIssuer =
                        VcCoreStatusIssuer(
                            sharedKms,
                            VcCoreFixtures.statusIssuerMetadata(keyMetadata),
                        )
                    (statusIssuer.issueStatusList(
                        VcCoreFixtures.STATUS_LIST_ID,
                        VcStatusesData.StatusListToken(listOf(StatusEntry(1u, 0u))),
                    ) as StatusList.StatusListTokenJwt)
                        .v1
                }
            mockServer.dispatcher =
                object : Dispatcher() {
                    override fun dispatch(request: RecordedRequest): MockResponse {
                        return if (request.path?.contains("status_list") == true) {
                            MockResponse()
                                .setResponseCode(200)
                                .setBody(statusListJwt)
                                .setHeader(
                                    "content-type",
                                    "application/statuslist+jwt",
                                )
                        } else {
                            MockResponse().setResponseCode(404)
                        }
                    }
                }
        }

        @JvmStatic
        @AfterAll
        fun tearDown() {
            mockServer.shutdown()
        }

        private fun newResolver(): UniversalDidResolver = UniversalDidResolver(null)
    }

    // -----------------------------------------------------------------
    // StatusIssuer
    // -----------------------------------------------------------------

    @Test
    fun statusIssuerIssuesStatusList() = runTest {
        val statusIssuer =
            VcCoreStatusIssuer(sharedKms, VcCoreFixtures.statusIssuerMetadata(keyMetadata))

        val result =
            statusIssuer.issueStatusList(
                VcCoreFixtures.STATUS_LIST_ID,
                VcStatusesData.StatusListToken(listOf(StatusEntry(1u, 0u))),
            )

        val jwt = (result as StatusList.StatusListTokenJwt).v1
        assertTrue(jwt.isNotEmpty(), "Expected non-empty JWT status list")
    }

    @Test
    fun statusIssuerBuilderProducesUsableInstance() = runTest {
        val builder =
            Oid4vciStatusIssuerBuilder(
                sharedKms,
                VcCoreFixtures.statusIssuerMetadata(keyMetadata),
            )
        val statusIssuer = builder.build()

        val result =
            statusIssuer.issueStatusList(
                VcCoreFixtures.STATUS_LIST_ID,
                VcStatusesData.StatusListToken(listOf(StatusEntry(2u, 1u))),
            )

        assertNotNull(result)
        assertTrue(result is StatusList.StatusListTokenJwt)
    }

    // -----------------------------------------------------------------
    // Issuer
    // -----------------------------------------------------------------

    @Test
    fun issuerOffersCredential() = runTest {
        val issuer =
            VcCoreIssuer(
                sharedKms,
                VcCoreFixtures.issuerMetadata(keyMetadata),
                newResolver(),
            )
        val offer = issuer.offerCredential(VcCoreFixtures.SCOPE, null)
        assertEquals(VcCoreFixtures.ISSUER_ID, offer.issuerId)
        assertEquals(VcCoreFixtures.SCOPE, offer.credDefId)
    }

    // -----------------------------------------------------------------
    // Holder
    // -----------------------------------------------------------------

    @Test
    fun holderRequestStoreAndVerifyCredential() = runTest {
        val vault = InMemVault()
        val resolver = newResolver()
        val issuer =
            VcCoreIssuer(
                sharedKms,
                VcCoreFixtures.issuerMetadata(keyMetadata),
                resolver,
            )
        val holder =
            VcCoreHolder(
                sharedKms,
                vault,
                VcCoreFixtures.holderMetadata(),
                resolver,
                ReqwestHttpClient.insecure(),
            )

        val offer = issuer.offerCredential(VcCoreFixtures.SCOPE, null)
        val credentialRequest =
            holder.requestCredential(offer, VcCoreFixtures.NONCE, keyMetadata)
        assertTrue(credentialRequest.proof.proof.isNotEmpty())

        val credential =
            issuer.issueCredential(
                credentialRequest,
                VcCoreFixtures.claims,
                VcCoreFixtures.NONCE,
                VcCoreFixtures.credStatusInfo,
            )
        holder.verifyCredential(credential)

        val metadata = resolveMetadata(credential, keyMetadata)
        val storedId = holder.storeCredential(credential, metadata)
        assertTrue(storedId.isNotEmpty())
    }

    @Test
    fun holderFindsAndPresentsCredential() = runTest {
        val vault = InMemVault()
        val resolver = newResolver()

        val issuer =
            VcCoreIssuer(
                sharedKms,
                VcCoreFixtures.issuerMetadata(keyMetadata),
                resolver,
            )
        val holder =
            VcCoreHolder(
                sharedKms,
                vault,
                VcCoreFixtures.holderMetadata(),
                resolver,
                ReqwestHttpClient.insecure(),
            )

        // Pre-issue + store one credential.
        val offer = issuer.offerCredential(VcCoreFixtures.SCOPE, null)
        val credentialRequest =
            holder.requestCredential(offer, VcCoreFixtures.NONCE, keyMetadata)
        val credential =
            issuer.issueCredential(
                credentialRequest,
                VcCoreFixtures.claims,
                VcCoreFixtures.NONCE,
                VcCoreFixtures.credStatusInfo,
            )
        holder.storeCredential(credential, resolveMetadata(credential, keyMetadata))

        // Find for presentation
        val findResult = holder.findVcsForPresentation(VcCoreFixtures.presentationInput())
        val data = findResult.data
        assertTrue(data is CredentialsSearchResult.Credentials, "expected credentials, got $data")
        assertTrue((data as CredentialsSearchResult.Credentials).v1.isNotEmpty())

        // Auto-present
        val binder =
            HolderBinder(
                nonce = VcCoreFixtures.NONCE,
                verifierId = VcCoreFixtures.VERIFIER_ID,
                responseUri = null,
            )
        val presentation =
            holder.createPresentationAuto(binder, VcCoreFixtures.presentationInput())
        assertTrue(presentation is Presentation.SdJwtVp, "Expected SdJwtVp presentation")
        assertTrue((presentation as Presentation.SdJwtVp).v1.isNotEmpty(), "Expected non-empty presentation payload")

        // Explicit-credential presentation
        val explicit =
            holder.createPresentation(
                binder,
                VcCoreFixtures.presentationInput(),
                data.v1.first(),
            )
        assertTrue(explicit is Presentation.SdJwtVp)
        assertTrue((explicit as Presentation.SdJwtVp).v1.isNotEmpty())
    }

    // -----------------------------------------------------------------
    // Verifier
    // -----------------------------------------------------------------

    @Test
    fun verifierVerifiesPresentation() = runTest {
        val vault = InMemVault()
        val resolver = newResolver()

        val issuer =
            VcCoreIssuer(
                sharedKms,
                VcCoreFixtures.issuerMetadata(keyMetadata),
                resolver,
            )
        val holder =
            VcCoreHolder(
                sharedKms,
                vault,
                VcCoreFixtures.holderMetadata(),
                resolver,
                ReqwestHttpClient.insecure(),
            )
        val verifier = VcCoreVerifier(VcCoreFixtures.VERIFIER_ID, resolver, ReqwestHttpClient.insecure())

        val offer = issuer.offerCredential(VcCoreFixtures.SCOPE, null)
        val credentialRequest =
            holder.requestCredential(offer, VcCoreFixtures.NONCE, keyMetadata)
        val credential =
            issuer.issueCredential(
                credentialRequest,
                VcCoreFixtures.claims,
                VcCoreFixtures.NONCE,
                VcCoreFixtures.credStatusInfo,
            )
        holder.storeCredential(credential, resolveMetadata(credential, keyMetadata))

        val binder =
            HolderBinder(
                nonce = VcCoreFixtures.NONCE,
                verifierId = VcCoreFixtures.VERIFIER_ID,
                responseUri = null,
            )
        val presentation =
            holder.createPresentationAuto(binder, VcCoreFixtures.presentationInput())

        val claimsJson =
            verifier.verifyPresentation(binder, presentation)
        assertTrue(
            claimsJson.contains("221B Baker Street"),
            "Expected verified claims to contain the address; got $claimsJson",
        )
        assertTrue(claimsJson.contains(VcCoreFixtures.VCT))
    }
}

File: wrappers/uniffi/swift/Tests/EqusSdkTests/VcCoreTest.swift
import Foundation
import Testing

@testable import EqusSdk

enum VcCoreFixtures {
    static let nonce = "KB50VOm9I-kPLT9mAACV8g"
    static let verifierId = "Verifier-id"
    static let scope = "SD_JWT_cred_sample"
    static let issuerId = "https://issuer-backend.com"
    static let statusListId = "test_status_list"
    static func statusListUrl(port: in_port_t) -> String { "http://localhost:\(port)/status_list" }
    static let vct = "https://credentials.example.com/identity_credential"

    static let claims: String = #"{"name":"John","surname":"Doe","address":"221B Baker Street","date":"09/09/1989"}"#

    static func credStatusInfo(port: in_port_t) -> CredentialStatusInfo {
        .tokenStatusList(idx: 1, uri: statusListUrl(port: port))
    }

    static func statusIssuerMetadata(keyMetadata: KeyMetadata, port: in_port_t) -> StatusIssuerMetadata {
        StatusIssuerMetadata(
            issuerId: "test",
            supportedStatusLists: [
                StatusListDefinition(
                    id: statusListId,
                    format: .statusListTokenJwt(
                        SlMetadata(
                            statusListUrl: statusListUrl(port: port),
                            statusesNr: 32,
                            statusSize: 2
                        )
                    ),
                    keyMetadata: keyMetadata
                )
            ]
        )
    }

    static func issuerMetadata(keyMetadata: KeyMetadata) -> IssuerMetadata {
        IssuerMetadata(
            issuerId: issuerId,
            credDefs: [
                CredentialDefinition(
                    credDefId: scope,
                    format: .sdJwtVc,
                    claims: [:],
                    supportedProofs: [SupportedProofEntry(format: .jwt, algs: [.es256])],
                    supportedSigningAlgs: [.es256, .edDsa],
                    display: nil,
                    protocolData: .sdJwt(
                        vct: vct,
                        disclosures: ["$.name", "$.surname"],
                        lifetime: 600
                    ),
                    keyMetadata: keyMetadata
                )
            ],
            protocolData: nil
        )
    }

    static func holderMetadata() -> HolderMetadata {
        let pop = ProofOfPossessionMetadataBuilder().withLifetime(lifetime: Int64(300)).build()
        return HolderMetadata(clientId: "wallet-dev", pop: pop)
    }

    static func presentationInput() -> PresentationInput {
        PresentationInput(
            id: scope,
            format: "dc+sd-jwt",
            restrictions: [
                PresentationRestriction(
                    fields: ["$.vct"],
                    value: .const(vct),
                    optional: false
                ),
                PresentationRestriction(
                    fields: ["$.surname"],
                    value: nil,
                    optional: false
                ),
            ]
        )
    }
}

@Suite(.serialized)
class VcCoreTests {
    let http: MockHttpRouter
    let port: in_port_t
    let kms: InMemKms
    let keyMetadata: KeyMetadata
    let statusListJwt: String

    init() async throws {
        self.http = MockHttpRouter()
        // No socket is bound; the router matches on path, so the port only has to
        // make the fixture URLs well-formed.
        self.port = 9000
        self.kms = InMemKms()
        let kvm = await createDidAndKeyMetadata(kms: kms)
        self.keyMetadata = kvm.keyMetadata

        let statusIssuer = try VcCoreStatusIssuer(
            kms: kms,
            metadata: VcCoreFixtures.statusIssuerMetadata(keyMetadata: keyMetadata, port: port)
        )
        let result = try await statusIssuer.issueStatusList(
            statusListId: VcCoreFixtures.statusListId,
            statuses: .statusListToken([StatusEntry(index: 1, status: 0)])
        )
        if case let .statusListTokenJwt(jwt) = result {
            self.statusListJwt = jwt
        } else {
            throw NSError(
                domain: "VcCoreTests", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "expected StatusListTokenJwt variant"]
            )
        }

        let cachedJwt = self.statusListJwt
        self.http["/status_list"] = { _ in
            MockHttpRouter.ok(cachedJwt, contentType: "application/statuslist+jwt")
        }
    }


    @Test func statusIssuerIssuesStatusList() async throws {
        let statusIssuer = try VcCoreStatusIssuer(
            kms: kms,
            metadata: VcCoreFixtures.statusIssuerMetadata(keyMetadata: keyMetadata, port: port)
        )
        let result = try await statusIssuer.issueStatusList(
            statusListId: VcCoreFixtures.statusListId,
            statuses: .statusListToken([StatusEntry(index: 1, status: 0)])
        )
        guard case let .statusListTokenJwt(jwt) = result else {
            throw NSError(domain: "VcCoreTests", code: 2, userInfo: [NSLocalizedDescriptionKey: "expected StatusListTokenJwt"])
        }
        #expect(!jwt.isEmpty)
    }

    @Test func statusIssuerBuilderProducesUsableInstance() async throws {
        let builder = Oid4vciStatusIssuerBuilder(
            kms: kms,
            metadata: VcCoreFixtures.statusIssuerMetadata(keyMetadata: keyMetadata, port: port)
        )
        let statusIssuer = try builder.build()
        let result = try await statusIssuer.issueStatusList(
            statusListId: VcCoreFixtures.statusListId,
            statuses: .statusListToken([StatusEntry(index: 2, status: 1)])
        )
        if case .statusListTokenJwt = result {
            // OK
        } else {
            throw NSError(domain: "VcCoreTests", code: 3, userInfo: [NSLocalizedDescriptionKey: "expected StatusListTokenJwt"])
        }
    }

    @Test func issuerOffersCredential() async throws {
        let resolver = try UniversalDidResolver(resolvers: nil)
        let issuer = try VcCoreIssuer(
            kms: kms,
            metadata: VcCoreFixtures.issuerMetadata(keyMetadata: keyMetadata),
            didResolver: resolver
        )
        let offer = try issuer.offerCredential(credDefId: VcCoreFixtures.scope, protocolData: nil)
        #expect(offer.issuerId == VcCoreFixtures.issuerId)
        #expect(offer.credDefId == VcCoreFixtures.scope)
    }

    @Test func holderRequestStoreAndVerifyCredential() async throws {
        let vault = InMemVault()
        let resolver = try UniversalDidResolver(resolvers: nil)
        let issuer = try VcCoreIssuer(
            kms: kms,
            metadata: VcCoreFixtures.issuerMetadata(keyMetadata: keyMetadata),
            didResolver: resolver
        )
        let holder = try VcCoreHolder(
            kms: kms,
            vault: vault,
            metadata: VcCoreFixtures.holderMetadata(),
            didResolver: resolver,
            httpClient: http
        )

        let offer = try issuer.offerCredential(credDefId: VcCoreFixtures.scope, protocolData: nil)
        let credentialRequest = try await holder.requestCredential(
            credentialOffer: offer,
            nonce: VcCoreFixtures.nonce,
            keyMetadata: keyMetadata
        )
        #expect(!credentialRequest.proof.proof.isEmpty)

        let credential = try await issuer.issueCredential(
            credentialRequest: credentialRequest,
            claims: VcCoreFixtures.claims,
            nonce: VcCoreFixtures.nonce,
            statusInfo: VcCoreFixtures.credStatusInfo(port: port)
        )
        try await holder.verifyCredential(credential: credential)

        let metadata = try await resolveMetadata(credential: credential, metadata: keyMetadata)
        let stored = try await holder.storeCredential(credential: credential, metadata: metadata)
        #expect(!stored.isEmpty)
    }

    @Test func holderFindsAndPresentsCredential() async throws {
        let vault = InMemVault()
        let resolver = try UniversalDidResolver(resolvers: nil)
        let issuer = try VcCoreIssuer(
            kms: kms,
            metadata: VcCoreFixtures.issuerMetadata(keyMetadata: keyMetadata),
            didResolver: resolver
        )
        let holder = try VcCoreHolder(
            kms: kms,
            vault: vault,
            metadata: VcCoreFixtures.holderMetadata(),
            didResolver: resolver,
            httpClient: http
        )

        let offer = try issuer.offerCredential(credDefId: VcCoreFixtures.scope, protocolData: nil)
        let credentialRequest = try await holder.requestCredential(
            credentialOffer: offer,
            nonce: VcCoreFixtures.nonce,
            keyMetadata: keyMetadata
        )
        let credential = try await issuer.issueCredential(
            credentialRequest: credentialRequest,
            claims: VcCoreFixtures.claims,
            nonce: VcCoreFixtures.nonce,
            statusInfo: VcCoreFixtures.credStatusInfo(port: port)
        )
        let metadata = try await resolveMetadata(credential: credential, metadata: keyMetadata)
        _ = try await holder.storeCredential(credential: credential, metadata: metadata)

        let findResult = try await holder.findVcsForPresentation(
            presentationInput: VcCoreFixtures.presentationInput()
        )
        guard case let .credentials(entries) = findResult.data else {
            throw NSError(
                domain: "VcCoreTests", code: 4,
                userInfo: [NSLocalizedDescriptionKey: "expected credentials, got \(findResult.data)"]
            )
        }
        #expect(!entries.isEmpty)

        let binder = HolderBinder(nonce: VcCoreFixtures.nonce, verifierId: VcCoreFixtures.verifierId, responseUri: nil)
        let presentation = try await holder.createPresentationAuto(
            holderBinder: binder,
            presentationInput: VcCoreFixtures.presentationInput()
        )
        guard case let .sdJwtVp(presentationPayload) = presentation else {
            throw NSError(domain: "VcCoreTests", code: 5, userInfo: [NSLocalizedDescriptionKey: "expected sdJwtVp presentation"])
        }
        #expect(!presentationPayload.isEmpty)

        let explicit = try await holder.createPresentation(
            holderBinder: binder,
            presentationInput: VcCoreFixtures.presentationInput(),
            credential: entries.first!
        )
        guard case let .sdJwtVp(explicitPayload) = explicit else {
            throw NSError(domain: "VcCoreTests", code: 6, userInfo: [NSLocalizedDescriptionKey: "expected sdJwtVp explicit presentation"])
        }
        #expect(!explicitPayload.isEmpty)
    }

    @Test func verifierVerifiesPresentation() async throws {
        let vault = InMemVault()
        let resolver = try UniversalDidResolver(resolvers: nil)
        let issuer = try VcCoreIssuer(
            kms: kms,
            metadata: VcCoreFixtures.issuerMetadata(keyMetadata: keyMetadata),
            didResolver: resolver
        )
        let holder = try VcCoreHolder(
            kms: kms,
            vault: vault,
            metadata: VcCoreFixtures.holderMetadata(),
            didResolver: resolver,
            httpClient: http
        )
        let verifier = try VcCoreVerifier(verifierId: VcCoreFixtures.verifierId, didResolver: resolver, httpClient: http)

        let offer = try issuer.offerCredential(credDefId: VcCoreFixtures.scope, protocolData: nil)
        let credentialRequest = try await holder.requestCredential(
            credentialOffer: offer,
            nonce: VcCoreFixtures.nonce,
            keyMetadata: keyMetadata
        )
        let credential = try await issuer.issueCredential(
            credentialRequest: credentialRequest,
            claims: VcCoreFixtures.claims,
            nonce: VcCoreFixtures.nonce,
            statusInfo: VcCoreFixtures.credStatusInfo(port: port)
        )
        let metadata = try await resolveMetadata(credential: credential, metadata: keyMetadata)
        _ = try await holder.storeCredential(credential: credential, metadata: metadata)

        let binder = HolderBinder(nonce: VcCoreFixtures.nonce, verifierId: VcCoreFixtures.verifierId, responseUri: nil)
        let presentation = try await holder.createPresentationAuto(
            holderBinder: binder,
            presentationInput: VcCoreFixtures.presentationInput()
        )

        let claimsJson = try await verifier.verifyPresentation(
            holderBinder: binder,
            presentation: presentation
        )
        #expect(claimsJson.contains("221B Baker Street"))
        #expect(claimsJson.contains(VcCoreFixtures.vct))
    }
}
